### DISCUSSION TITLE
controlling the draw order of osgEarth LabelNodes.

### DISCUSSION THREAD

**Discussion by Apadona** [pelicanmapping/osgearth#2422] [Support Q&A] [answered]
Hello dear OsgEarth community.
i have a problem regarding the drawing of osgEarth::Annotation::LabelNodes. and that is that LabelNodes always gets drawn over all other objects.
like here:

![problem](https://github.com/gwaldron/osgearth/assets/85191978/2ef3cd81-de15-4bba-9b8e-8c6d46790ff2)

as seen, the label 'object 1' is drawn on cube and it is overwriting its pixels.

i have tried various functions such as label->setOcclusionCulling(true) and label->setPriority(negative_or_positive_value); or even tried to change its RenderBinDetails from its osg::StateSet to a big number such as label->getOrCreateStateSet()->setRenderBinDetails(999, "RenderBin"), but all it does is that it makes it not being displayed.

how can i controll the draw order of the LabelNode such that other objects would always occulude if they were to occupy the same pixels.

any help would be greatly appreciated.

**Comment by gwaldron** [comment, marked as answer]
LabelNodes are not like normal scene objects. They draw in a screen-space overlay after the rest of the scene renders. The only way to get them to "interact" with the scene objects is by using the occlusion culling option. This option intersects a ray from the camera to the label's position, and if it hits anything, it skips drawing the label.

The occlusion culling also has a global max altitude at which is takes effect -- perhaps you need to increase this by calling `AnnotationSettings::setOcclusionCullingMaxAltitude(...)` -- the default value is 200000m.

If you want complete 3D sorting and rendering for text, you will need to instead use osgEarth::Text objects directly in your code.

Hope this helps.

**Reply by Apadona** [reply]
thank you, it was helpful.
i've tried the setOcclussionCullingMaxAltitude function, but it still doesn't work. the version of the osgEarth that we are using (2.9.0) is older than the current available version. so i believe i have to implement my own custom Text class based on osgText::Text and some sorting/drawing implementations so that it would function like the LabelNodes.
regardless, i thank you again for your quick response on the matter.
